pymongo/asynchronous/helpers.py

```diff
@@ -0,0 +1,317 @@
+
+
+def _check_write_command_response(result: Mapping[str, Any]) -> None:
+    """Backward compatibility helper for write command error handling."""
+    # Prefer write errors over write concern errors
+    write_errors = result.get("writeErrors")
+    if write_errors:
+        _raise_last_write_error(write_errors)
+
+    wce = _get_wce_doc(result)
+    if wce:
+        _raise_write_concern_error(wce)
+
+
+def _fields_list_to_dict(
+    fields: Union[Mapping[str, Any], Iterable[str]], option_name: str
+) -> Mapping[str, Any]:
+    """Takes a sequence of field names and returns a matching dictionary.
+
+    ["a", "b"] becomes {"a": 1, "b": 1}
+
+    and
+
+    ["a.b.c", "d", "a.c"] becomes {"a.b.c": 1, "d": 1, "a.c": 1}
+    """
+    if isinstance(fields, abc.Mapping):
+        return fields
+
+    if isinstance(fields, (abc.Sequence, abc.Set)):
+        if not all(isinstance(field, str) for field in fields):
+            raise TypeError(f"{option_name} must be a list of key names, each an instance of str")
+        return dict.fromkeys(fields, 1)
+
+    raise TypeError(f"{option_name} must be a mapping or list of key names")
+
+
+def _handle_exception() -> None:
+    """Print exceptions raised by subscribers to stderr."""
+    # Heavily influenced by logging.Handler.handleError.
+
+    # See note here:
+    # https://docs.python.org/3.4/library/sys.html#sys.__stderr__
+    if sys.stderr:
+        einfo = sys.exc_info()
+        try:
+            traceback.print_exception(einfo[0], einfo[1], einfo[2], None, sys.stderr)
+        except OSError:
+            pass
+        finally:
+            del einfo
+
+
+# See https://mypy.readthedocs.io/en/stable/generics.html?#decorator-factories
+F = TypeVar("F", bound=Callable[..., Any])
+
+
+def _handle_reauth(func: F) -> F:
+    async def inner(*args: Any, **kwargs: Any) -> Any:
+        no_reauth = kwargs.pop("no_reauth", False)
+        from pymongo.asynchronous.message import _BulkWriteContext
+        from pymongo.asynchronous.pool import Connection
+
+        try:
+            return func(*args, **kwargs)
+        except OperationFailure as exc:
+            if no_reauth:
+                raise
+            if exc.code == _REAUTHENTICATION_REQUIRED_CODE:
+                # Look for an argument that either is a Connection
+                # or has a connection attribute, so we can trigger
+                # a reauth.
+                conn = None
+                for arg in args:
+                    if isinstance(arg, Connection):
+                        conn = arg
+                        break
+                    if isinstance(arg, _BulkWriteContext):
+                        conn = arg.conn
+                        break
+                if conn:
+                    await conn.authenticate(reauthenticate=True)
+                else:
+                    raise
+                return func(*args, **kwargs)
+            raise
+
+    return cast(F, inner)
+
+
+async def anext(cls: Any) -> Any:
```

Comment:
`anext` was added in Python 3.10, so we need to use our own version here for compatibility with Python 3.8 and 3.9. `anext` is one of the special names that `unasync` automatically converts to the sync version, which is `next` in this case.

Reply:
I suggest using the builtin method on 3.10+ and only using the fallback on <3.9.